datafusion/functions/src/strings.rs

```diff
@@ -124,8 +125,18 @@ pub struct StringArrayBuilder {
 
 impl StringArrayBuilder {
     pub fn with_capacity(item_capacity: usize, data_capacity: usize) -> Self {
-        let mut offsets_buffer =
-            MutableBuffer::with_capacity((item_capacity + 1) * size_of::<i32>());
+        // MutableBuffer::with_capacity will round up to multiples of `ALIGNMENT`
+        let capacity = std::cmp::max(
+            item_capacity.saturating_add(1).saturating_mul(ALIGNMENT),
+            data_capacity,
+        );
+        // rust core's Layout::from_size_align is capped at isize::Max
+        let capacity = std::cmp::min(
```

Comment:
I am not sure what case you are trying to prevent by checking `ALIGNMENT`, but it looks like maybe you have found a problem passing large (but not overflowing) values to `MutableBuffer::with_capacity`

If this is the case, I have two suggestions:
1. Such a check belongs in arrow-rs (`MutableBuffer::with_capacity`)
2. File a ticket explaining what the issue is in arrow-rs along with a reproducer before adding additional checks.